// expo-cli/build/commands/build-native/iOSBuilder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _buildTools() {
  const data = require("@expo/build-tools");

  _buildTools = function () {
    return data;
  };

  return data;
}

function _iOSCredentialsProvider() {
  const data = _interopRequireDefault(require("../../credentials/provider/iOSCredentialsProvider"));

  _iOSCredentialsProvider = function () {
    return data;
  };

  return data;
}

function _credentials() {
  const data = require("./credentials");

  _credentials = function () {
    return data;
  };

  return data;
}

function _easJson() {
  const data = require("../../easJson");

  _easJson = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class iOSBuilder {
  constructor(ctx) {
    this.ctx = ctx;

    _defineProperty(this, "credentials", void 0);

    _defineProperty(this, "buildProfile", void 0);

    if (!ctx.eas.builds.ios) {
      throw new Error("missing ios configuration, shouldn't happen");
    }

    this.buildProfile = ctx.eas.builds.ios;
  }

  async prepareJobAsync(archiveUrl) {
    if (this.buildProfile.workflow === _easJson().Workflow.Generic) {
      return (0, _buildTools().sanitizeJob)((await this.prepareGenericJobAsync(archiveUrl, this.buildProfile)));
    } else if (this.buildProfile.workflow === _easJson().Workflow.Managed) {
      return (0, _buildTools().sanitizeJob)((await this.prepareManagedJobAsync(archiveUrl, this.buildProfile)));
    } else {
      throw new Error("Unknown workflow. Shouldn't happen");
    }
  }

  async ensureCredentialsAsync() {
    var _this$ctx$exp, _this$ctx$exp$ios;

    if (!this.shouldLoadCredentials()) {
      return;
    }

    const bundleIdentifier = (_this$ctx$exp = this.ctx.exp) === null || _this$ctx$exp === void 0 ? void 0 : (_this$ctx$exp$ios = _this$ctx$exp.ios) === null || _this$ctx$exp$ios === void 0 ? void 0 : _this$ctx$exp$ios.bundleIdentifier;

    if (!bundleIdentifier) {
      throw new Error('"expo.ios.bundleIdentifier" field is required in your app.json');
    }

    const provider = new (_iOSCredentialsProvider().default)(this.ctx.projectDir, {
      projectName: this.ctx.projectName,
      accountName: this.ctx.accountName,
      bundleIdentifier
    });
    await provider.initAsync();
    const credentialsSource = await (0, _credentials().ensureCredentialsAsync)(provider, this.buildProfile.workflow, this.buildProfile.credentialsSource);
    this.credentials = await provider.getCredentialsAsync(credentialsSource);
  }

  async prepareJobCommonAsync(archiveUrl) {
    const secrets = this.credentials ? {
      secrets: {
        provisioningProfileBase64: this.credentials.provisioningProfile,
        distributionCertificate: {
          dataBase64: this.credentials.distributionCertificate.certP12,
          password: this.credentials.distributionCertificate.certPassword
        }
      }
    } : {};
    return {
      platform: _buildTools().Platform.iOS,
      projectUrl: archiveUrl,
      ...secrets
    };
  }

  async prepareGenericJobAsync(archiveUrl, buildProfile) {
    return { ...(await this.prepareJobCommonAsync(archiveUrl)),
      type: _buildTools().BuildType.Generic
    };
  }

  async prepareManagedJobAsync(archiveUrl, buildProfile) {
    return { ...(await this.prepareJobCommonAsync(archiveUrl)),
      type: _buildTools().BuildType.Managed,
      packageJson: {
        example: 'packageJson'
      },
      manifest: {
        example: 'manifest'
      }
    };
  }

  shouldLoadCredentials() {
    return this.buildProfile.workflow === _easJson().Workflow.Managed && this.buildProfile.buildType !== 'simulator' || this.buildProfile.workflow === _easJson().Workflow.Generic;
  }

}

var _default = iOSBuilder;
exports.default = _default;
